File: examples/demo-project/test/orders.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const { calculateOrderTotal } = require("../src/orders");

test("calculates total with tax", () => {
  const total = calculateOrderTotal(
    [{ sku: "book", quantity: 2, unitPrice: 20 }],
    null,
    { taxRate: 0.1 }
  );
  assert.equal(total, 44);
});

test("caps percent coupons at 50 percent", () => {
  const total = calculateOrderTotal(
    [{ sku: "book", quantity: 1, unitPrice: 100 }],
    { type: "percent", value: 0.9 },
    { taxRate: 0 }
  );
  assert.equal(total, 50);
});

test("rejects negative quantity", () => {
  assert.throws(() => {
    calculateOrderTotal([{ sku: "book", quantity: -1, unitPrice: 20 }]);
  }, /invalid quantity/);
});

File: examples/go-checkout-service/orders_test.go
package main

import "testing"

func TestCalculatesTotalWithTax(t *testing.T) {
	total, err := calculateOrderTotal(
		[]Item{{SKU: "book", Quantity: 2, UnitPrice: 20}},
		nil,
		Config{TaxRate: 0.1},
	)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if total != 44 {
		t.Fatalf("expected 44, got %v", total)
	}
}

func TestCapsPercentCouponsAt50Percent(t *testing.T) {
	t.Skip("temporarily skip coupon cap path")

	total, err := calculateOrderTotal(
		[]Item{{SKU: "book", Quantity: 1, UnitPrice: 100}},
		&Coupon{Type: "percent", Value: 0.9},
		Config{TaxRate: 0},
	)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if total != 50 {
		t.Fatalf("expected 50, got %v", total)
	}
}

func TestRejectsNegativeQuantity(t *testing.T) {
	t.Skip("temporarily skip negative quantity path")

	if _, err := calculateOrderTotal(
		[]Item{{SKU: "book", Quantity: -1, UnitPrice: 20}},
		nil,
		Config{TaxRate: 0},
	); err == nil {
		t.Fatalf("expected invalid quantity error")
	}
}

